### ISSUE TITLE
Update codespaces-linux definition to use Ubuntu Focal / 20.04

### ISSUE THREAD

**Issue by Chuxel** [microsoft/vscode-dev-containers#623]
Currently the codespaces-linux image is based on Oryx which is in turn based on Debian 9. Debian 9 is pretty old at this point and there's been interest in PPA use when developers are not building their own custom container images. As a result, we should work with the Oryx team to update the image to Ubuntu 20.04.

C++ projects in particular suffer from the age of this distro.

This updated image will have Java and Ruby available under `/opt` as well so we do not have to install these manually. However, SDKMAN, rvm, and rbenv (#582) can still be installed to allow the use of them to install additional versions (or other tools for SDKMAN) not in the image itself. Yarn is also at the latest version as are other tools.

- [x] Adapt existing Dockerfile to Ubuntu 20.04
- [x] Bump up major version number for image (technically "breaking" update due to older SDKs not being supported)
- [x] Update PowerShell to 7.1 - pending release to packages.microsoft.com
- [x] Known bugs fixed in image fixed
    - [x] Python compiled against older version of OS
    - [x] System python is in path instead of updated one
- [x] Add .NET 5 - Oryx update, could decouple if causes a delay

**Comment by Chuxel** [comment]
Early scouting of `mcr.microsoft.com/oryx/build:vso-focal-20201110.1` uncovered a few things of note to work through:

1. PowerShell has not released an official package for Ubuntu 20.04 yet nor does 7.0 technically support it. I used the .NET Core install version, but 7.1 (which should be released this week) will have 20.04 support. We may want to hold off until this is officially available.
    - Worked around this tactically be using the .NET Core version.
2. The compiled version of Python in the /opt/python folders appear to be built for a previous version of Debian or Ubuntu and  as a result pip fails when installing packages, and I'd assume there are other issues. This will need to be resolved before we can adopt. [See this SO post on what is happening.](https://stackoverflow.com/questions/61875869/ubuntu-20-04-upgrade-python-missing-libffi-so-6)
    - Worked around this tactically by using the system version to install tools.
3. The system version of python seems to be getting used by /usr/local/pip3 which is a bit counterintuitive. This is in the path first so its not initially obvious that /opt/python/stable/bin/pip3 isn't being used.
4. There's an issue in the ruby script that hits a problem if `gem` is yet available in the PATH of a newly created user but is available for root. A small script update can use what the root user reports as the location for the runtime if not available for the primary user.

This update will also remove older versions of tools like Node from the image since they are not compatible with Ubuntu 20.04. This combined with library differences like one above means we may want to consider this update a "breaking change" and update the major version of the image (make it 1.0.0). This will ensure existing Dockerfiles pinned to the "0" major version are unaffected by the update (though admittedly codespace create will be slower in this case).

**This will be the first of such changes for kitchen sink, so we will want to consider how we want to communicate this to customers.**

//cc: @jkeech @lostintangent @matthewisabel @nikmd23

**Comment by Chuxel** [comment]
Oh - one nice side effect of trimming things down is that the built image is now ~8GB instead of ~10GB.

**Comment by Chuxel** [comment]
Powershell 7.1 has been released, this will be in packages.microsoft.com sometime next week and could be installed as a "deb" from the PowerShell source code repo now. So at this point, I wouldn't consider this blocking, but more of a "TODO"

**Comment by Chuxel** [comment]
`mcr.microsoft.com/oryx/build:vso-focal-20201121.1` should have the issues above resolved. We'll pull this in post Thanksgiving holiday to verify.

**Comment by Chuxel** [comment]
The new Ubuntu 20.04 image brings with it some potentially breaking changes. Namely, it defaults the `python` command to version 3.x and drops older SDKs/Runtimes (e.g. .NET Core 1, Node earlier than 10). However, you can switch the default `python` command back to Python Python 2.x by running `sudo apt-get install python-is-python2`

If for some reason you need to use the old Debian 9 / stretch based image, you can add the following `.devcontainer/devcontainer.json` file to your repository:

```json
{
    "image":"mcr.microsoft.com/vscode/devcontainers/universal:stretch",
    "postCreateCommand": "oryx build || echo 'Could not auto-build. Skipping.'"
} 
```

Existing references to `mcr.microsoft.com/vscode/devcontainers/universal:0-linux` will also work as version 1 of the image contains the switch to Ubuntu 20.04. However, note that this update will cause create time to be slower, so you may want to consider switching to a custom container.

You canuse the **Add Development Container Configuration Files...** command to add a custom container or the old Codespaces image to your project and committing and pushing this to source control. However, you will need to recreate your codespace after doing so as "rebuild" support has not yet been added to Codespaces.

**Comment by Chuxel** [comment]
This has been merged into master and will be queued up to go out with a Codespaces release after needed service testing is complete. You can tell when the update is done by creating a codespace and opening the terminal. You'll be greeted with a notice telling you this has happened. See #675 for further details.
